src/EoS/Nuclear/CPU_EoS_Nuclear.cpp

```diff
@@ -454,18 +481,266 @@ static real EoS_DensPres2CSqr_Nuclear( const real Dens_Code, const real Pres_Cod
+//                         In [1] = entropy in kB/baryon
+//                         In [2] = Ye
+//                         Out[0] = volume energy density in code units (with energy shift)
+//                   --> Both In[] and Out[] can be extended if necessary
+//                4. No units conversion will be applied to entropy and Ye
+//
+// Parameter   :  Mode       : NUC_MODE_TEMP or NUC_MODE_ENTR
+//                Out        : Output array     (see the Note above)
+//                In         : Input array      (see the Note above)
+//                AuxArray_* : Auxiliary arrays (see the Note above)
+//                Table      : EoS tables
+//
+// Return      :  Out[]
+//-------------------------------------------------------------------------------------------------------
+GPU_DEVICE_NOINLINE
+static void EoS_General_Nuclear( const int Mode, real Out[], const real In[], const double AuxArray_Flt[],
+                                 const int AuxArray_Int[], const real *const Table[EOS_NTABLE_MAX] )
+{
+
+// general check
+#  ifdef GAMER_DEBUG
+   if ( Out          == NULL )   printf( "ERROR : Out == NULL in %s !!\n", __FUNCTION__ );
+   if ( In           == NULL )   printf( "ERROR : In == NULL in %s !!\n", __FUNCTION__ );
+   if ( AuxArray_Flt == NULL )   printf( "ERROR : AuxArray_Flt == NULL in %s !!\n", __FUNCTION__ );
+   if ( AuxArray_Int == NULL )   printf( "ERROR : AuxArray_Int == NULL in %s !!\n", __FUNCTION__ );
+#  endif // GAMER_DEBUG
+
+
+   switch ( Mode )
+   {
+//    temperature mode
+      case NUC_MODE_TEMP :
+      {
+         const real Dens_Code = In[0];
+         const real Temp_Kelv = In[1];
+         const real Ye        = In[2];
+
+//       check
+#        ifdef GAMER_DEBUG
+         if ( Hydro_CheckNegative(Dens_Code) )
+            printf( "ERROR : invalid input density (%14.7e) at file <%s>, line <%d>, function <%s>\n",
+                    Dens_Code, __FILE__, __LINE__, __FUNCTION__ );
+
+         if ( Hydro_CheckNegative(Temp_Kelv) )
+            printf( "ERROR : invalid input temperature (%14.7e) at file <%s>, line <%d>, function <%s>\n",
+                    Temp_Kelv, __FILE__, __LINE__, __FUNCTION__ );
+
+         if ( Hydro_CheckNegative(Ye) )
+            printf( "ERROR : invalid input Ye (%14.7e) at file <%s>, line <%d>, function <%s>\n",
+                    Ye, __FILE__, __LINE__, __FUNCTION__ );
+#        endif // GAMER_DEBUG
+
+         const real EnergyShift = AuxArray_Flt[NUC_AUX_ESHIFT    ];
+         const real Dens2CGS    = AuxArray_Flt[NUC_AUX_DENS2CGS  ];
+         const real Kelvin2MeV  = AuxArray_Flt[NUC_AUX_KELVIN2MEV];
+         const real sEint2Code  = AuxArray_Flt[NUC_AUX_VSQR2CODE ];
+
+         const int  NRho        = AuxArray_Int[NUC_AUX_NRHO ];
+         const int  NEps        = AuxArray_Int[NUC_AUX_NEPS ];
+         const int  NYe         = AuxArray_Int[NUC_AUX_NYE  ];
+         const int  NMode       = AuxArray_Int[NUC_AUX_NMODE];
+
```

Comment:
How about move these assignments (`AuxArray_Flt` and `AuxArray_Int`) before the switch statement since both modes require them (one more `Kelvin2MeV` in the temperature mode)?